In [ ]:
import numpy as np

In [ ]:
np.__version__

'1.25.2'

Vamos começar instanciando o mesmo array do módulo anterior, que representa o desempenho de 5 atletas em 4 voltas de uma pista de corrida, em 3 dias diferentes.

In [ ]:
rng = np.random.default_rng(seed=1)

In [ ]:
array = rng.normal(loc=50, scale=3, size=60).reshape(3, 5, 4).round(3)

In [ ]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [ ]:
array.shape

(3, 5, 4)

👆 Observe pelo `shape` que são 3 dias, 5 atletas e 4 voltas.

### Métodos de conversão

#### `item`

Utilize o método `item` para recuperar o desempenho do segundo atleta no terceiro dia, na primeira volta.

In [ ]:
...

49.705

#### `itemset`

Suponha que este valor tenha sido registrado errado, e que o valor correto seja 49.709. Faça a correção utilizando `itemset`, e exiba o array corrigido.

In [ ]:
...

In [ ]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.709, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Utilize `item` de novo para ter certeza de que o valor foi alterado.

In [ ]:
...

49.709

#### `view`

Gere uma view deste array, atribuindo o resultado a um novo array `array_view`.

In [ ]:
array_view = ...

Utilize `itemset` para mudar o valor corrigido anteriormente (índice (2, 1, 0)) para seu valor original (49.705), apenas em `array_view`.

In [ ]:
...

Agora, utilize `item` para verificar o valor no índice (2, 1, 0) em `array` e `array_view`.

In [ ]:
array....

49.705

In [ ]:
array_view....

49.705

💡 Observe que, apesar de você ter feito a modificação em `array_view`, `array` também foi modificado. Isso acontece porque, sendo `array_view` uma **view** de `array`, seus dados são compartilhados.

#### `copy`

Agora repita o exercício anterior, mas criando uma cópia de `array`, armazenada em `array_copy`.

In [ ]:
array_copy = ...

Mude de novo o valor em (2, 1, 0) para 49.709, somente na cópia.

In [ ]:
...

Verifique com `view` que este valor agora é diferente em `array` e `array_copy`.

In [ ]:
array....

49.705

In [ ]:
array_copy....

49.709

💡 Uma **cópia** de um array mantém dados independentes do array original.

### Métodos de manipulação de `shape`

#### `reshape`

Suponha que você precise registrar os resultados do desempenho dos atletas em uma planilha, que armazena dados em apenas 2 dimensões. A orientação que você recebeu é que cada linha da planilha deve ser correspondente a um dia, totalizando 3 linhas, e as colunas vão conter as 4 voltas do primeiro atleta, depois as 4 voltas do segundo atleta, e assim por diante, totalizando 5 atletas x 4 voltas = 20 colunas.

Exiba uma versão do array nesta configuração.

In [ ]:
...

array([[51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
        51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
        50.119, 49.123, 47.654, 49.228],
       [50.024, 49.173, 53.882, 53.02 , 41.867, 44.333, 49.476, 48.733,
        50.641, 50.652, 56.354, 46.664, 48.867, 56.128, 51.94 , 51.989,
        48.458, 45.056, 50.502, 50.327],
       [46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
        51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
        52.743, 49.94 , 46.254, 49.058]])

#### `transpose`

Lembre-se que as dimensões do array representam os dias, os atletas e as voltas, nesta ordem.

Utilize `transpose` para representar o array com os dados dos atletas, depois as voltas, depois os dias. Armazene o resultado em um novo `array_transpose`, exiba o array e, depois seu `shape`.

In [ ]:
array_transpose = ...

In [ ]:
array_transpose

array([[[51.037, 50.024, 46.318],
        [52.465, 49.173, 47.95 ],
        [50.991, 53.882, 49.784],
        [46.091, 53.02 , 47.166]],

       [[52.716, 41.867, 49.705],
        [51.339, 44.333, 50.286],
        [48.389, 49.476, 50.107],
        [51.743, 48.733, 48.481]],

       [[51.094, 50.641, 51.781],
        [50.882, 50.652, 52.674],
        [50.085, 56.354, 50.963],
        [51.64 , 46.664, 47.545]],

       [[47.791, 48.867, 52.195],
        [49.511, 56.128, 48.496],
        [48.554, 51.94 , 52.637],
        [51.797, 51.989, 46.785]],

       [[50.119, 48.458, 52.743],
        [49.123, 45.056, 49.94 ],
        [47.654, 50.502, 46.254],
        [49.228, 50.327, 49.058]]])

In [ ]:
...

(5, 4, 3)

👆 5 atletas, 4 voltas, 3 dias.

💡 Ou seja, a primeira linha da visualização:
> ```
> [51.037, 50.024, 46.318]
> ```

representa o primeiro atleta, a primeira volta, e os 3 dias.

#### `swapaxes`

A partir de `array_transpose`, agora utilize `swapaxes` para mudar os eixos de forma a representar os atletas, depois os dias, depois as voltas. Armazene o resultado em `array_swapaxes`, exiba o array e seu shape.

In [ ]:
array_swapaxes = ...

In [ ]:
array_swapaxes

array([[[51.037, 52.465, 50.991, 46.091],
        [50.024, 49.173, 53.882, 53.02 ],
        [46.318, 47.95 , 49.784, 47.166]],

       [[52.716, 51.339, 48.389, 51.743],
        [41.867, 44.333, 49.476, 48.733],
        [49.705, 50.286, 50.107, 48.481]],

       [[51.094, 50.882, 50.085, 51.64 ],
        [50.641, 50.652, 56.354, 46.664],
        [51.781, 52.674, 50.963, 47.545]],

       [[47.791, 49.511, 48.554, 51.797],
        [48.867, 56.128, 51.94 , 51.989],
        [52.195, 48.496, 52.637, 46.785]],

       [[50.119, 49.123, 47.654, 49.228],
        [48.458, 45.056, 50.502, 50.327],
        [52.743, 49.94 , 46.254, 49.058]]])

In [ ]:
...

(5, 3, 4)

👆 5 atletas, 3 dias, 4 voltas.

💡 Ou seja, a primeira linha da visualização:
> ```
> [51.037, 52.465, 50.991, 46.091]
> ```

representa o primeiro atleta, o primeiro dia, as 4 voltas.

#### `flatten` / `ravel`

Agora suponha que você tenha que registrar os resultados em um sistema que foi programado para receber dados de forma linear, um valor por vez.

Linearize o array utilizando `flatten` e `ravel`.

In [ ]:
...

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354, 46.664,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
       46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

In [ ]:
...

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354, 46.664,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
       46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

💡 Lembre-se que `flatten` retorna uma **cópia** do array, enquanto que `ravel` retorna uma **view**, que compartilha os dados do `array`, mas mantém seu próprio `shape`.

#### `squeeze`

Nós podemos separar os resultados do primeiro dia em um novo array `array_first_day`, com a seguinte notação:

In [ ]:
array_first_day = ...

In [ ]:
array_first_day

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]]])

In [ ]:
array_first_day.ndim, array_first_day.shape

(3, (1, 5, 4))

Observe que o número de dimensões (`ndim`) de `array_first_day` continua sendo igual a 3, ou seja, o mesmo de `array`. Com `shape`, nós observamos que o eixo correspondente aos dias possui um único elemento, justamente o primeiro dia selecionado.

Utilize `squeeze` para simplificar este array, removendo o eixo dos dias, já que ele contém somente um único dia. Salve o resultado em `array_squeeze`, exiba o array e, depois, seu `shape`.

In [ ]:
array_squeeze = ...

In [ ]:
array_squeeze

array([[51.037, 52.465, 50.991, 46.091],
       [52.716, 51.339, 48.389, 51.743],
       [51.094, 50.882, 50.085, 51.64 ],
       [47.791, 49.511, 48.554, 51.797],
       [50.119, 49.123, 47.654, 49.228]])

👆 Veja que tem um par de colchetes a menos, o que indica a redução de 1 eixo.

In [ ]:
...

(5, 4)

👆 O eixo correspondente ao dia, de tamanho 1, foi removido.

### Métodos de seleção e manipulação de itens

#### `take`

Agora considere que os dados em `array` se referem ao mesmo atleta. O primeiro eixo continua representando 3 dias diferentes, o último representa as 4 voltas de uma prova, mas o segundo eixo representa 5 provas completas do mesmo atleta. Ou seja, `array` agora contém dados consecutivos do mesmo atleta.

Suponha que, assim que terminou as provas, o atleta se queixou de dor no joelho nas voltas de número 12, 25, 31 e 40. Utilizando `take`, recupere o desempenho do atleta nessas voltas.

In [ ]:
...

array([47.791, 44.333, 46.664, 46.318])

Mas o atleta foi mais específico e disse que, nas voltas 12 e 31, a dor foi no joelho direito, e nas voltas 25 e 40, a dor foi no joelho esquerdo. Utilizando `take`, recupere essas voltas novamente, mas de forma a obter os resultados como um array onde a primeira linha se refira ao joelho direito, e a segunda ao joelho esquerdo.

In [ ]:
...

array([[47.791, 46.664],
       [44.333, 46.318]])

#### `put`

Com base no que o atleta reportou, seu treinador decidiu desconsiderar o desempenho do atleta nas voltas onde ele se queixou de dor. Utilize `put` para mudar estes valores para a constante `np.nan`, que representa valores nulos em um array do NumPy. Depois, exiba o array modificado.

In [ ]:
...

In [ ]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

#### `repeat`

Repita os valores do array para que eles sejam duplicados (`repeat = 2`), ao longo do eixo 0, 1 e 2, separadamente.

Depois, exiba o array para inspecionar onde ocorreu a repetição. Exiba também o `shape` do array para observar a alteração feita.

In [ ]:
array_repeat_axis_0 = ...

In [ ]:
array_repeat_axis_0

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 5

In [ ]:
...

(6, 5, 4)

👆 Veja que os dados correspondentes aos dias foram duplicados.

In [ ]:
array_repeat_axis_1 = ...

In [ ]:
array_repeat_axis_1

array([[[51.037, 52.465, 50.991, 46.091],
        [51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [49.705, 50.286, 50.10

In [ ]:
...

(3, 10, 4)

👆 Veja que os dados correspondentes às provas foram duplicados.

In [ ]:
array_repeat_axis_2 = ...

In [ ]:
array_repeat_axis_2

array([[[51.037, 51.037, 52.465, 52.465, 50.991, 50.991, 46.091, 46.091],
        [52.716, 52.716, 51.339, 51.339, 48.389, 48.389, 51.743, 51.743],
        [51.094, 51.094, 50.882, 50.882, 50.085, 50.085, 51.64 , 51.64 ],
        [   nan,    nan, 49.511, 49.511, 48.554, 48.554, 51.797, 51.797],
        [50.119, 50.119, 49.123, 49.123, 47.654, 47.654, 49.228, 49.228]],

       [[50.024, 50.024, 49.173, 49.173, 53.882, 53.882, 53.02 , 53.02 ],
        [41.867, 41.867,    nan,    nan, 49.476, 49.476, 48.733, 48.733],
        [50.641, 50.641, 50.652, 50.652, 56.354, 56.354,    nan,    nan],
        [48.867, 48.867, 56.128, 56.128, 51.94 , 51.94 , 51.989, 51.989],
        [48.458, 48.458, 45.056, 45.056, 50.502, 50.502, 50.327, 50.327]],

       [[   nan,    nan, 47.95 , 47.95 , 49.784, 49.784, 47.166, 47.166],
        [49.705, 49.705, 50.286, 50.286, 50.107, 50.107, 48.481, 48.481],
        [51.781, 51.781, 52.674, 52.674, 50.963, 50.963, 47.545, 47.545],
        [52.195, 52.195, 48.496, 4

In [ ]:
...

(3, 5, 8)

👆 Veja que os dados correspondentes às voltas foram duplicados.

#### `sort`

Primeiro, crie uma cópia de `array`, já que `sort` altera o array *inplace*.

Depois, utilize `sort` para ordenar o array ao longo do eixo 0. Exiba o array ordenado para observar o que aconteceu.

In [ ]:
array_copy = ...

In [ ]:
array_copy....

In [ ]:
array_copy

array([[[50.024, 47.95 , 49.784, 46.091],
        [41.867, 50.286, 48.389, 48.481],
        [50.641, 50.652, 50.085, 47.545],
        [48.867, 48.496, 48.554, 46.785],
        [48.458, 45.056, 46.254, 49.058]],

       [[51.037, 49.173, 50.991, 47.166],
        [49.705, 51.339, 49.476, 48.733],
        [51.094, 50.882, 50.963, 51.64 ],
        [52.195, 49.511, 51.94 , 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[   nan, 52.465, 53.882, 53.02 ],
        [52.716,    nan, 50.107, 51.743],
        [51.781, 52.674, 56.354,    nan],
        [   nan, 56.128, 52.637, 51.989],
        [52.743, 49.94 , 50.502, 50.327]]])

👆 Observe que, ao longo dos dias, os valores estão ordenados de forma crescente: `50.024 -> 51.037 -> nan`, `47.95 -> 49.173 -> 52.465`...

Também observe que o NumPy coloca `nan` no **final** da ordem; todos os valores `nan` estão no terceiro elemento do eixo 0.

Agora repita o procedimento, copiando o array e depois ordenando ao longo do eixo 1.

In [ ]:
array_copy = ...

In [ ]:
...

In [ ]:
array_copy

array([[[50.119, 49.123, 47.654, 46.091],
        [51.037, 49.511, 48.389, 49.228],
        [51.094, 50.882, 48.554, 51.64 ],
        [52.716, 51.339, 50.085, 51.743],
        [   nan, 52.465, 50.991, 51.797]],

       [[41.867, 45.056, 49.476, 48.733],
        [48.458, 49.173, 50.502, 50.327],
        [48.867, 50.652, 51.94 , 51.989],
        [50.024, 56.128, 53.882, 53.02 ],
        [50.641,    nan, 56.354,    nan]],

       [[49.705, 47.95 , 46.254, 46.785],
        [51.781, 48.496, 49.784, 47.166],
        [52.195, 49.94 , 50.107, 47.545],
        [52.743, 50.286, 50.963, 48.481],
        [   nan, 52.674, 52.637, 49.058]]])

👆 Agora, a ordenação ocorre ao longo das provas (que na visualização correspondem às colunas): `50.119 -> 51.037 -> 51.094 -> 52.716 -> nan`...

Desta vez, os valores `nan` estão no último elemento do eixo 1 (última linha de cada array que representa os dias).

Repita mais uma vez, copiando o array e depois ordenando ao longo do eixo 2.

In [ ]:
array_copy = ...

In [ ]:
...

In [ ]:
array_copy

array([[[46.091, 50.991, 51.037, 52.465],
        [48.389, 51.339, 51.743, 52.716],
        [50.085, 50.882, 51.094, 51.64 ],
        [48.554, 49.511, 51.797,    nan],
        [47.654, 49.123, 49.228, 50.119]],

       [[49.173, 50.024, 53.02 , 53.882],
        [41.867, 48.733, 49.476,    nan],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 51.94 , 51.989, 56.128],
        [45.056, 48.458, 50.327, 50.502]],

       [[47.166, 47.95 , 49.784,    nan],
        [48.481, 49.705, 50.107, 50.286],
        [47.545, 50.963, 51.781, 52.674],
        [46.785, 48.496, 52.195, 52.637],
        [46.254, 49.058, 49.94 , 52.743]]])

👆 A ordenação ocorre ao longo das voltas (que na visualização correspondem às linhas): `46.091 -> 50.991 -> 51.037 -> 52.465`...

Desta vez, os valores `nan` estão no último elemento do eixo 2 (última coluna  de cada array que representa os dias).

#### `argsort`

Repita o exercício com `sort`, mas agora, ao invés de ordenar os valores *inplace*, recupere e exiba os índices necessários para ordenar o array. Isto é feito com o método `argsort`.

Desta vez, como não há modificação do array, você não precisa da cópia.

In [ ]:
indices_axis_0 = ...

In [ ]:
indices_axis_0

array([[[1, 2, 2, 0],
        [1, 2, 0, 2],
        [1, 1, 0, 2],
        [1, 2, 0, 2],
        [1, 1, 2, 2]],

       [[0, 1, 0, 2],
        [2, 0, 1, 1],
        [0, 0, 2, 0],
        [2, 0, 1, 0],
        [0, 0, 0, 0]],

       [[2, 0, 1, 1],
        [0, 1, 2, 0],
        [2, 2, 1, 1],
        [0, 1, 2, 1],
        [2, 2, 1, 1]]])

👆 Observe que o valor máximo neste array que representa os índices é 2, porque o eixo 0 só contém 3 elementos (3 dias).

In [ ]:
indices_axis_1 = ...

In [ ]:
indices_axis_1

array([[[4, 4, 4, 0],
        [0, 3, 1, 4],
        [2, 2, 3, 2],
        [1, 1, 2, 1],
        [3, 0, 0, 3]],

       [[1, 4, 1, 1],
        [4, 0, 4, 4],
        [3, 2, 3, 3],
        [0, 3, 0, 0],
        [2, 1, 2, 2]],

       [[1, 0, 4, 3],
        [2, 3, 0, 0],
        [3, 4, 1, 2],
        [4, 1, 2, 1],
        [0, 2, 3, 4]]])

👆 Agora o valor máximo é 4, pois o eixo 1 tem 5 elementos (5 atletas).

In [ ]:
indices_axis_2 = ...

In [ ]:
indices_axis_2

array([[[3, 2, 0, 1],
        [2, 1, 3, 0],
        [2, 1, 0, 3],
        [2, 1, 3, 0],
        [2, 1, 3, 0]],

       [[1, 0, 3, 2],
        [0, 3, 2, 1],
        [0, 1, 2, 3],
        [0, 2, 3, 1],
        [1, 0, 3, 2]],

       [[3, 1, 2, 0],
        [3, 0, 2, 1],
        [3, 2, 0, 1],
        [3, 1, 0, 2],
        [2, 3, 1, 0]]])

👆 O valor máximo é 3, pois o eixo 2 tem 4 elementos (4 voltas).

#### `searchsorted`

Este método só funciona para arrays 1D que já estejam ordenados.

Então, primeiro linearize `array` com `flatten`, depois ordene com `sort` (lembre-se que o método `sort` altera o array *inplace*, ou seja, ele não retorna nada, então você não precisa armazenar o resultado em uma nova variável), e utilize indexação para recuperar os primeiros 15 elementos.

In [ ]:
array_flatten = ...

In [ ]:
...

In [ ]:
array_filtered = ...

In [ ]:
array_filtered

array([41.867, 45.056, 46.091, 46.254, 46.785, 47.166, 47.545, 47.654,
       47.95 , 48.389, 48.458, 48.481, 48.496, 48.554, 48.733])

Utilizando `searchsorted`, descubra em que posições deveriam ser inseridos os valores 40.50, 42.357, 47.025 e 48.952 deveriam ser inseridos para manter a ordenação.

In [ ]:
indices = ...

In [ ]:
indices

array([ 0,  1,  5, 15])

👆 Inspecione o array e os índices encontrados para confirmar a resposta.

#### `diagonal`

O entendimento de diagonais é facilitado trabalhando com arrays 2D.

Por isso, primeiro obtenha com indexação uma view do array correspondente ao primeiro dia.

In [ ]:
array_2D = ...

In [ ]:
array_2D

array([[51.037, 52.465, 50.991, 46.091],
       [52.716, 51.339, 48.389, 51.743],
       [51.094, 50.882, 50.085, 51.64 ],
       [   nan, 49.511, 48.554, 51.797],
       [50.119, 49.123, 47.654, 49.228]])

In [ ]:
array_2D.shape

(5, 4)

Agora, recupere e exiba a diagonal deste array.

In [ ]:
...

array([51.037, 51.339, 50.085, 51.797])

👆 Não é difícil encontrar visualmente esta diagonal em `array_2D`.

Observe que, como `array_2D` é **retangular** (como demonstra seu `shape`), nenhum valor da última linha é retornado na diagonal, pois ele tem mais linhas que colunas.